### PR TITLE
Serialise communication with controller to reduce connection errors

### PR DESCRIPTION
**Pull request by callifo**
Had some issues for a while where after a few days the controller connection would become unstable, it was evident from wireshark captures the controller would only allow a number of concurrent connections to it. Initially after bootup it was three, but this would decrease over time (possibly related to restarting HA). Eventually no connection could be sustained and a controller reboot was required.

The get_resource calls and send_command calls are overlapping and eventually they are not responded to. Even though getting the first SystemSettings call is sucessfull, the Zones1_4 and/or Zones5_8 fail and this cause the controller to go unavailable. 

I've modified it to move the sending lock, now controller lock, directly to the lowest level calls accessing the controller (get_resource, send_command_async). I've run this code now for more than a week, and I haven't had a single connection drop the whole time. Previously I had had issues where after reboot HA, it would discover the controller but remain unavailable for a few minutes or five, this also no longer is an issue.

Fresh controller boot, overlapping calls
tcpdump

After a day, some succeed, others attempt TCP retransmit but ultimately fail. One one call is complete, the next succeeds.
tcpdump_1

After these changes, no overlaps,
tcpdump_2

### CODE REVIEW

File: pizone/zone.py [hunk 1]
```diff
@@ -107,15 +107,14 @@ async def set_airflow_min(self, value: int) -> None:
...
-            await self._send_command('AirMinCommand', value)
+        await self._send_command('AirMinCommand', value)
 
-            # need to refresh immediatley after updating
-            try:
-                await self._controller._refresh_zone_group(
-                    self._index - self._index % 4)
-            except ConnectionError:
-                pass
+        # need to refresh immediatley after updating
```
Comment (Swamp-Ig) [inline pizone/zone.py:112]: I think these might need to be inside the sending lock to avoid a race condition.

**Comment by Swamp-Ig** [comment]
Sorry for the late reply. Does the server refresh need to be inside the sending lock? I think it might need to.

**Comment by callifo** [comment]
Can you suggest where the race condition might be? I'm not great with python, but I've placed the controller_lock(s) at the lowest level where the actual request is made to the izone controller as I had issues placing them any higher due to deadlocks (taking the lock, and then encountering a get/send request that then needs the same lock).

The set_airflow_min method in  zone.py calls _refresh_zone_group in controller.py which calls _get_resource, which _get_resource method contains the controller lock when it sends to the izone controller. Placing it around _refresh_zone_group and it then later being locked when it gets to _get_resource might cause a deadlock?

I have had this running on mine since I submitted the PR with debugging enabled and haven't noticed any issues.

**Comment by Swamp-Ig** [comment]
The concern was that the controller can only handle one request at a time, they all have to be serialised. If the ```_get_resource``` happens concurrently (as in two requests at the same time) I suspect it may crash the controller.

Could be wrong though, it's probs more an issue with updating than getting.

**Comment by callifo** [comment]
Ok, so this is the part of concern,

    async def _get_resource(self, resource: str):
        try:
            session = self._discovery.session
            async with self._controller_lock:
                async with session.get(
                        'http://%s/%s' % (self.device_ip, resource),
                        timeout=Controller.REQUEST_TIMEOUT) as response:
                    return await response.json(content_type=None)
        except (asyncio.TimeoutError, aiohttp.ClientError) as ex:
            self._failed_connection(ex)
            raise ConnectionError("Unable to connect to the controller") \
                from ex

If _get_resource is called twice concurrently, will the 2nd line, "async with self._controller_lock:" not cause the second call to wait until the first request & lock is completed?

Looking the inline comment above, are you suggesting the lock needs to be applied like this?

        try:
            async with self._controller_lock:
                await self._controller._refresh_zone_group(
                    self._index - self._index % 4)
        except ConnectionError:
            pass

If so, I can run it up on mine for a bit to see how it goes.

**Comment by Swamp-Ig** [comment]
Yes you're correct, it would. The only issue is that you go outside the lock between the set and the refresh, which means another set might happen before the refresh. I guess it won't matter too much though, since there's always going to be a refresh at the end of every set anyhow so if you end up serializing as set_1 set_2 refresh_2 refresh_1, but the delayed refresh_1 just gets the latest state anyhow.

**Comment by Swamp-Ig** [comment]
I do believe the locks are reentrant anyhow, so if the current thread already owns the lock then locking it again works regardless. Might be worth testing.

**Comment by callifo** [comment]
Seems to be ok with that change, will run it for a few days and then update the PR.

**Comment by callifo** [comment]
Silly me, didnt actually read what method this was apart of, the set_min_airflow and set_max_airflow are services in HA. I have to admit I've didn't know they existed and I have therefore never used them before. 

I tried running them without the changes and things worked as expected, airflow_min/max went up/down fine. 

With the lock included in the _set_airflow_min/max, I found I was unable to run either of these services more than the first time. It would change the value for the zone, however, it would kill any ability to use that service or any further izone functions until I reboot HA. It seems to be locking up the whole izone climate device.

I think as I mentioned above, its getting stuck, because we use a lock in the _set_min/max_airflow, and then again when calling _send_command_async its already taken and the whole module now gets stuck; any further _get_resource or _send_command_async can't run until the lock is released.

Any suggestions?

**Comment by Swamp-Ig** [comment]
Must be deadlocking. I guess the locks aren't re-entrant. 

I was going to have a look at doing some coding tomorrow, so will try and have a poke.

**Comment by Swamp-Ig** [comment]
I'm thinking actually what I'll do is whenever there's an update I'll fire off a settings poll in the poll thread, and block any property reading until this returns.  That way all communication is serialised and there's only one in flight at a time.

**Comment by Swamp-Ig** [comment]
I think I might have fixed the issue. The results weren't being read back from the connection all the way to EOF, which potentially caused a comms back-log. I have fixed this now in something I've just pushed up. Will start testing it anyhow.

**Comment by callifo** [comment]
I only follow part of what your saying :) 

Do you need me to do anything for the time being?

**Comment by Swamp-Ig** [comment]
Yeh, don't worry, am working on it. I have discovered another issue in the controller which I'll have to work around.

**Comment by Swamp-Ig** [comment]
@califo  I have updated the code now, do you want to try the current head and see how it goes? Published as version v1.1.7

**Comment by Swamp-Ig** [comment]
Closing this, have updated the codebase

**Comment by callifo** [comment]
@Swamp-Ig Sorry I missed this, I just given 1.1.8 a test and it seems good. I can reboot multiple times and it doesn't produce errors and stop working. Thanks

**Comment by Swamp-Ig** [comment]
I've published 1.2.0 too. This handles another connection bug better without losing connection. Give it a try if you like. 

I'm currently working on getting the power monitoring included into HA,

**Pull request closed without merging**